filebeat/input/file/state.go

```diff
@@ -18,82 +18,93 @@
 	"github.com/elastic/beats/v7/libbeat/common/file"
 )
 
 // State is used to communicate the reading state of a file
 type State struct {
-	Id          string            `json:"-"` // local unique id to make comparison more efficient
-	Finished    bool              `json:"-"` // harvester state
-	Fileinfo    os.FileInfo       `json:"-"` // the file info
-	Source      string            `json:"source"`
-	Offset      int64             `json:"offset"`
-	Timestamp   time.Time         `json:"timestamp"`
-	TTL         time.Duration     `json:"ttl"`
-	Type        string            `json:"type"`
-	Meta        map[string]string `json:"meta"`
-	FileStateOS file.StateOS
+	Id             string            `json:"-"` // local unique id to make comparison more efficient
+	Finished       bool              `json:"-"` // harvester state
+	Fileinfo       os.FileInfo       `json:"-"` // the file info
+	Source         string            `json:"source"`
+	Offset         int64             `json:"offset"`
+	Timestamp      time.Time         `json:"timestamp"`
+	TTL            time.Duration     `json:"ttl"`
+	Type           string            `json:"type"`
+	Meta           map[string]string `json:"meta"`
+	Data           interface{}       `json:"data,omitempty"` // arbitrary data of a comparator
+	ComparatorName string            `json:"comparator_name"`
+	FileStateOS    file.StateOS
+
+	identifier StateIdentifier
 }
 
 // NewState creates a new file state
-func NewState(fileInfo os.FileInfo, path string, t string, meta map[string]string) State {
+func NewState(fileInfo os.FileInfo, path string, t string, meta map[string]string, comparator IdentityComparator) State {
 	if len(meta) == 0 {
 		meta = nil
 	}
+
+	identifier := comparator.StateIdentifier()
+
 	return State{
-		Fileinfo:    fileInfo,
-		Source:      path,
-		Finished:    false,
-		FileStateOS: file.GetOSState(fileInfo),
-		Timestamp:   time.Now(),
-		TTL:         -1, // By default, state does have an infinite ttl
-		Type:        t,
-		Meta:        meta,
+		Fileinfo:       fileInfo,
+		Source:         path,
+		Finished:       false,
+		FileStateOS:    file.GetOSState(fileInfo),
+		Timestamp:      time.Now(),
+		TTL:            -1, // By default, state does have an infinite ttl
+		Type:           t,
+		Meta:           meta,
+		ComparatorName: comparator.Name(),
+		Data:           identifier.DataToSerialize(),
+		identifier:     identifier,
 	}
 }
 
-// ID returns a unique id for the state as a string
+// ID returns the ID of the state.
 func (s *State) ID() string {
-	// Generate id on first request. This is needed as id is not set when converting back from json
-	if s.Id == "" {
-		if len(s.Meta) == 0 {
-			s.Id = s.FileStateOS.String()
-		} else {
-			hashValue, _ := hashstructure.Hash(s.Meta, nil)
-			var hashBuf [17]byte
-			hash := strconv.AppendUint(hashBuf[:0], hashValue, 16)
-			hash = append(hash, '-')
-
-			fileID := s.FileStateOS.String()
+	return s.identifier.ID(s)
```

Comment:
Ah I see, identifier creates the `ID` lazily and caches the result. This is a 'bad' habit from the past and actually used to lead to performance problems in the past (this is why we introduced caching via `Id`). The ID is used almost always for looking up the state in the registry, do a comparison, garbage collecting stale entries, or to check if the file/state is already known. Maybe we want to create the ID eagerly always. WDYT?